synapse/handlers/message.py

```diff
@@ -915,41 +936,44 @@ async def create_new_client_event(
                 Should normally be left as None, which will cause them to be calculated
                 based on the room state at the prev_events.
 
+            full_state_ids_at_event:
+                The event ids for the full state at that event. This is used
+                particularly by the MSC2716 /batch_send endpoint which shares the same
+                state across the whole batch. The state will be stripped down to only
+                what's necessary for the auth_event_ids. For insertion events, we will
+                add this as the explicit state so the rest of the histroical batch can
+                inherit the same state and state_group.
+
             depth: Override the depth used to order the event in the DAG.
                 Should normally be set to None, which will cause the depth to be calculated
                 based on the prev_events.
 
         Returns:
             Tuple of created event, context
         """
-        # Strip down the auth_event_ids to only what we need to auth the event.
+        # Strip down the full_state_ids_at_event to only what we need to auth the event.
         # For example, we don't need extra m.room.member that don't match event.sender
-        full_state_ids_at_event = None
-        if auth_event_ids is not None:
-            # If auth events are provided, prev events must be also.
-            # prev_event_ids could be an empty array though.
-            assert prev_event_ids is not None
-
-            # Copy the full auth state before it stripped down
-            full_state_ids_at_event = auth_event_ids.copy()
-
+        if full_state_ids_at_event is not None:
             temp_event = await builder.build(
                 prev_event_ids=prev_event_ids,
-                auth_event_ids=auth_event_ids,
+                auth_event_ids=full_state_ids_at_event,
                 depth=depth,
             )
-            auth_events = await self.store.get_events_as_list(auth_event_ids)
+            state_events = await self.store.get_events_as_list(full_state_ids_at_event)
             # Create a StateMap[str]
-            auth_event_state_map = {
-                (e.type, e.state_key): e.event_id for e in auth_events
-            }
-            # Actually strip down and use the necessary auth events
+            state_map = {(e.type, e.state_key): e.event_id for e in state_events}
+            # Actually strip down and only use the necessary auth events
             auth_event_ids = self._event_auth_handler.compute_auth_events(
                 event=temp_event,
-                current_state_ids=auth_event_state_map,
+                current_state_ids=state_map,
                 for_verification=False,
             )
 
+        if auth_event_ids is not None:
+            # If auth events are provided, prev events must be also.
+            # prev_event_ids could be an empty array though.
+            assert prev_event_ids is not None
```

Comment:
De-nesting this check since the `if-statement` above turned into `if state_event_ids is not None:`